src/Utilities/SingleStackUtils/SingleStackUtils.jl

```diff
@@ -298,6 +299,35 @@ function reduce_element_stack(
     ]
 end
 
+"""
+    horizontally_average!(
+        grid::DiscontinuousSpectralElementGrid{T, dim, N},
+        Q::MPIStateArray,
+        i_vars,
+    ) where {T, dim, N}
+
+Horizontally average variables, from variable
+indexes `i_vars`, in `MPIStateArray` `Q`.
+"""
+function horizontally_average!(
+    grid::DiscontinuousSpectralElementGrid{T, dim, N},
+    Q::MPIStateArray,
+    i_vars,
+) where {T, dim, N}
+    Nq = N + 1
+    Nqk = dimensionality(grid) == 2 ? 1 : Nq
+    for ev in 1:size(Q, 3), k in 1:Nqk, i_v in i_vars
+        Q_sum = sum((1:Nq, 1:Nq)) do ij
+            Q[ij[1] + Nq * ((ij[2] - 1) + Nq * (k - 1)), i_v, ev]
+        end
+        Q_ave = Q_sum / (Nq * Nq)
+        for i in 1:Nq, j in 1:Nq
+            ijk = i + Nq * ((j - 1) + Nq * (k - 1))
+            Q[ijk, i_v, ev] = Q_ave
+        end
+    end
+end
```

Comment:
This should probably use the metric terms to compute the average, mainly you should compute the integral and divide by the length. Otherwise, you are giving more weight to the solution near the boundary due to the use of LGL points.

Reply:
Also, as written this is not a GPU friendly function.

Reply:
None of the single stack util functions are GPU friendly. They started as just a couple of helper functions and are now growing rapidly -- probably need refactoring to abstract out MPIStateArray internal knowledge and for GPU friendliness soonish now.

Reply:
> This should probably use the metric terms to compute the average, mainly you should compute the integral and divide by the length. Otherwise, you are giving more weight to the solution near the boundary due to the use of LGL points.

Thanks for letting me know. The main purpose of this function is to eliminate horizontal variation introduced in `rand()`, used in `init_state_prognostic!` for some of the single stack drivers. Ultimately, I'm trying to eliminate any resulting horizontal fluxes-- do I need to compute this average using the metric terms in order to eliminate horizontal fluxes?

> Also, as written this is not a GPU friendly function.

This is not performance-critical code (it's just for initialization), but we should modify it to work when the backing arrays are CuArrays.

Reply:
> Ultimately, I'm trying to eliminate any resulting horizontal fluxes-- do I need to compute this average using the metric terms in order to eliminate horizontal fluxes?

OK. I misunderstood what you were doing. In this case, you are right that it doesn't matter too much.

Also, I pointed to the wrong code snippet, you don't want the vertical integral but the horizontal integral stuff that was put into the diagnostics: https://github.com/CliMA/ClimateMachine.jl/blob/fa76456047bba885a17f9f73756ed3e4b2cd6eba/src/Diagnostics/atmos_les_core.jl#L65

Reply:
Ok, it should now work on the GPU.